FindFFTW.cmake

```diff
@@ -201,139 +218,228 @@ else()
-               INTERFACE_LINK_LIBRARIES "${FFTW_DOUBLETHREADS_LIB}"
-  )
-else()
-  set(FFTW_DOUBLE_THREADS_LIB_FOUND FALSE)
-endif()
+if ( FFTW_USE_MKL )
 
-if (FFTW_FLOAT_THREADS_LIB)
-  set(FFTW_FLOAT_THREADS_LIB_FOUND TRUE)
-  set(FFTW_LIBRARIES ${FFTW_LIBRARIES} ${FFTW_FLOAT_THREADS_LIB})
-  add_library(FFTW::FloatThreads INTERFACE IMPORTED)
-  set_target_properties(FFTW::FloatThreads
-    PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${FFTW_INCLUDE_DIRS}"
-               INTERFACE_LINK_LIBRARIES "${FFTW_FLOAT_THREADS_LIB}"
-  )
-else()
-  set(FFTW_FLOAT_THREADS_LIB_FOUND FALSE)
-endif()
+  # Set FFTW include directories
+  set(FFTW_INC_SEARCHPATH ${FFTW_ROOT}/include/fftw)
 
-if (FFTW_LONGDOUBLE_THREADS_LIB)
-  set(FFTW_LONGDOUBLE_THREADS_LIB_FOUND TRUE)
-  set(FFTW_LIBRARIES ${FFTW_LIBRARIES} ${FFTW_LONGDOUBLE_THREADS_LIB})
-  add_library(FFTW::LongDoubleThreads INTERFACE IMPORTED)
-  set_target_properties(FFTW::LongDoubleThreads
-    PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${FFTW_INCLUDE_DIRS}"
-               INTERFACE_LINK_LIBRARIES "${FFTW_LONGDOUBLE_THREADS_LIB}"
+  find_path(FFTW_INCLUDE_PATH
+    NAMES "fftw3.h"
+    PATHS ${FFTW_INC_SEARCHPATH}
+    NO_DEFAULT_PATH # so it finds the fftw.h file in the MKL directory
   )
-else()
-  set(FFTW_LONGDOUBLE_THREADS_LIB_FOUND FALSE)
-endif()
 
-if (FFTW_DOUBLE_OPENMP_LIB)
-  set(FFTW_DOUBLE_OPENMP_LIB_FOUND TRUE)
-  set(FFTW_LIBRARIES ${FFTW_LIBRARIES} ${FFTW_DOUBLE_OPENMP_LIB})
-  add_library(FFTW::DoubleOpenMP INTERFACE IMPORTED)
-  set_target_properties(FFTW::DoubleOpenMP
-    PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${FFTW_INCLUDE_DIRS}"
-               INTERFACE_LINK_LIBRARIES "${FFTW_DOUBLE_OPENMP_LIB}"
-  )
-else()
-  set(FFTW_DOUBLE_OPENMP_LIB_FOUND FALSE)
-endif()
+  if( FFTW_INCLUDE_PATH )
+    file(TO_CMAKE_PATH "${FFTW_INCLUDE_PATH}" FFTW_INCLUDE_PATH)
+    set(FFTW_INCLUDE_DIRS ${FFTW_INCLUDE_PATH})
+  endif()
+
+  # Create list of FFTW libraries to find
+  if(CMAKE_SIZEOF_VOID_P EQUAL "8")
+    # 64-bit OS
+    if(APPLE)
+      set(FFTW_LIB_SEARCHPATH ${FFTW_ROOT}/lib)
+    else()
+      set(FFTW_LIB_SEARCHPATH ${FFTW_ROOT}/lib/intel64)
+    endif()
+    set(MKL_LIBRARY mkl_intel_ilp64)
+    if(WIN32)
+      set(MKL_OPTIONS /DMKL_ILP64)
+    else()
+      set(MKL_OPTIONS -DMKL_ILP64 -m64)
+    endif()
+  
+  else()
+    # not 64-bit OS
+    set(FFTW_LIB_SEARCHPATH ${FFTW_ROOT}/lib/ia32)
+    if(WIN)
+      set(MKL_LIBRARY mkl_intel_c)
+    else()
+      set(MKL_LIBRARY mkl_intel)
+    endif()
+  endif()
+
+  set(MKL_EXTRA_LIBRARIES mkl_core)
+  list(APPEND MKL_EXTRA_LIBRARIES mkl_sequential)
+  set(FFTW_LIBRARY_NAMES ${MKL_LIBRARY} ${MKL_EXTRA_LIBRARIES})
+  
+  macro(FFTWD_LIB_START)
+    unset(FFTWD_LIB CACHE)
+    unset(FFTWF_LIB CACHE)
+    unset(FFTWD_THREADS_LIB CACHE)
+    unset(FFTWF_THREADS_LIB CACHE)
+    if(${CMAKE_SYSTEM_NAME} MATCHES "Linux")
+      set(FFTWD_LIB -Wl,--start-group)
+    endif()
+  endmacro()
```

Comment:
What do you need this for? If you want to set compile options, I would certainly not do that inside this `find_package` module. A better way would be to set an "output" variable (list it above and in the readme) like `FFTW_MKL_LINKER_OPTIONS` or something like that (seems like these are linker options, right?). Then users can set these options on their targets themselves after their `find_package` call.

Reply:
It uses 64-bit ints to index into large arrays with more than 2^31-1 elements (https://software.intel.com/en-us/mkl-macos-developer-guide-using-the-ilp64-interface-vs-lp64-interface). I don't know if this is the best file for it since it's not FFTW-specific, so I'm removing it.